packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/GeoPolygon.kt

```diff
@@ -0,0 +1,123 @@
+ * argument for a geospatial query.
+ *
+ * A [GeoPolygon] describes a shape compromised of 3 or more line segments. It comprises of one
+ * outer ring and 0 or more rings representing holes with the following restrictions:
+ *
+ * - Each ring must consist of at least 3 distinct points (vertices). The first and the last point
+ *   must be the same to indicate a closed ring (meaning you need at least 4 points to define the
+ *   polygon).
+ * - Rings may not cross, i.e. the boundary of a ring may not intersect both the interior and
+ *   exterior of any other ring.
+ * - Rings may not share edges, i.e. if a ring contains an edge AB, then no other ring may contain
+ *   AB or BA.
+ * - Rings may share vertices, however no vertex may appear twice in a single ring.
+ * - No ring may be empty.
+ *
+ * Holes may be nested inside each other, in which case a location will be considered "inside" the
+ * polygon if it is included in an odd number of rings. For example, a polygon representing a square
+ * with side 10 centered at (0,0) with holes representing squares with sides 5 and 2, centered at
+ * (0,0) will  include the location (1, 1) because it is contained in 3 rings, but not (3, 3),
+ * because it is contained in 2.
+ *
+ * Using it can look something like this:
+ *
+ * ```
+ * val newYork = GeoPoint(latitude = 40.730610, longitude = -73.935242)
+ * val searchArea = GeoCircle(center = newYork, radius = Distance.fromMiles(2.0))
+ * val restaurants = realm.query<Restaurant>("location GEOWITHIN $searchArea").find()
+ * ```
+ */
+public data class GeoPolygon(
+
+    /**
+     * A list of [GeoPoint]s representing the outer ring of the polygon.
+     *
+     * The outer ring must contain at least 3 unique points. The first and the last point may
+     * be identical, but no other duplicates are allowed. Each subsequent pair of points represents
+     * an edge in the polygon with the first and the last point being implicitly connected.
+     */
+    public val outerRing: List<GeoPoint>,
+
+    /**
+     * A list of "inner" rings. Each of these rings has the same requirements as the [outerRing].
+     *
+     * A point is considered "inside" the polygon if it is contained by an odd number of rings
+     * and "outside" if it's contained by an even number of rings.
+     */
+    public val holes: Iterable<List<GeoPoint>> = emptyList()
+) {
+
+    /**
+     * Construct a polygon with a single inner ring. This will create a polygon that only matches
+     * locations between these two rings.
+     */
+    public constructor(outerRing: List<GeoPoint>, hole: List<GeoPoint>) : this(outerRing, listOf(hole))
+
+    private companion object {
+        private const val MIN_RING_SIZE = 3
+    }
+
+    init {
+        // Do basic input validation. Core will validate the rest when the query is run.
+        // Better input validation will be available later.
+        if (outerRing.size < MIN_RING_SIZE) {
```

Comment:
Should we check that the first and the last point is the same. You have some tests where it isn't the case. I think it would be a nice guide for users.

Reply:
Yes, I guess that is an easy check to add. Core also detects it, but then it doesn't happen until you actually run the query.